File: csharp/Ice/Filesystem/Server/MDirectory.cs
// Copyright (c) ZeroC, Inc.

namespace Server;

/// <summary>Provides an in-memory implementation of the Slice interface Directory.</summary>
internal class MDirectory : Filesystem.DirectoryDisp_
{
    private readonly MNode _node; // reuse the MNode implementation
    private readonly List<Filesystem.NodePrx> _contents = [];

    /// <inheritdoc/>
    // Implements Node.Name by forwarding to _node.
    public override string Name(Ice.Current current) => _node.Name(current);

    /// <inheritdoc/>
    public override Filesystem.NodePrx[] List(Ice.Current current) => _contents.ToArray();

    /// <summary>Constructs a MDirectory.</summary>
    /// <param name="name">The name of this directory.</param>
    internal MDirectory(string name) => _node = new MNode(name);

    /// <summary>Adds a node to this directory.</summary>
    /// <param name="child">The node proxy to add.</param>
    internal void AddChild(Filesystem.NodePrx child) => _contents.Add(child);
}

File: python/Ice/filesystem/server/m_directory.py
# Copyright (c) ZeroC, Inc.

import Ice
from Filesystem import Directory, NodePrx
from m_node import MNode


class MDirectory(Directory, MNode):
    """
    Provides an in-memory implementation of the Slice interface Directory.
    """

    def __init__(self, name: str):
        """
        Initializes an MDirectory with the given name.

        Parameters
        ----------
        name : str
            The name of the directory.
        """
        super().__init__(name)
        self._contents = []

    def list(self, current: Ice.Current) -> list[NodePrx]:
        return self._contents

    def addChild(self, child: NodePrx) -> None:
        """
        Adds a node to this directory.

        Parameters
        ----------
        child : NodePrx
            The node to add.
        """
        self._contents.append(child)
